support for writing X509 requests in DER format

afcady opened the pull request:
I guess you don't need this patch for your own project, but it probably belongs in your pull request to the upstream.

Actually, it occurred to me when I was looking at the OpenSSL.PEM module that the DER functions logically belong in (a new module named) OpenSSL.DER.  Perhaps you would be willing to put them all there?

photm5 commented:
Sorry for the long delay :(

> I guess you don't need this patch for your own project, but it probably
> belongs in your pull request to the upstream.

I’m not sure about that – From what I can tell, there’s a PR for DER in X509
already:  https://github.com/phonohawk/HsOpenSSL/pull/39

I generally don’t like to include other people’s code in my PRs when they’re
just adding to it and not fixing my mistakes.  What I added is useful on its own
outside of X509.

It seems you have added some more bits than the PR referenced above, so you
might want to sort out what capabilities you can add and open a seperate PR at
https://github.com/phonohawk/HsOpenSSL that is based on that other PR.

> Actually, it occurred to me when I was looking at the OpenSSL.PEM module that
> the DER functions logically belong in (a new module named) OpenSSL.DER.
> Perhaps you would be willing to put them all there?

Ah thanks, I think you’re right there.  I’ll update my PR soon and then you can
base your PR on it, adding the X509 things to that new module.

Anyways, thanks for reminding me – I planned to improve my PR (By adding support
for private keys as well), but always postponed it.

afcady commented:
> From what I can tell, there’s a PR for DER in X509 already: phonohawk#39 

That code is already merged into my PR here -- see https://github.com/afcady/HsOpenSSL/commit/9c2f05c3e70f6e93a6e9a0ece4b2f48606a0c158).  (I  did that so I could change from String to ByteString, because decoding the binary DER data into a String can throw an exception if the string is not valid unicode).

But note that that PR is actually about X509 _certificates_ whereas the code that I needed (and wrote myself, as opposed to merely merging) was for X509 certificate _requests_.

(Also, at the time that I made this PR, I had not yet merged @newsham 's code, so maybe you saw it back then.)

> I generally don’t like to include other people’s code in my PRs when they’re just adding to it and not fixing my mistakes. What I added is useful on its own outside of X509. 

Fair enough.  I was just hoping to relinquish responsibility as easily as possible for myself, of course.  It does not look like HsOpenSSL devs are responsive to PRs, so that I was hoping someone else -- already committed to dealing with them -- would take on the baton and I could forget about it.

Anyway, do note, this PR to you does incorporate all of the DER functionality in both of the PRs (yours and @newsham 's) plus what I added.  Of course you should do as you please, but to me it seems logical that if one should do any reorganization of modules, it would be based on this PR.

afcady commented:
I went ahead and made PR https://github.com/phonohawk/HsOpenSSL/pull/49
